Objects/typeobject.c

```diff
@@ -3420,88 +3449,89 @@ PyType_FromMetaclass(PyTypeObject *metaclass, PyObject *module,
     if (spec->name == NULL) {
         PyErr_SetString(PyExc_SystemError,
                         "Type spec does not define the name field.");
-        goto fail;
+        goto finally;
     }
 
-    /* Adjust for empty tuple bases */
+    /* Get a tuple of bases.
+     * bases is a strong reference (unlike bases_in).
+     */
+    bases = get_bases_tuple(bases_in, spec);
     if (!bases) {
-        base = &PyBaseObject_Type;
-        /* See whether Py_tp_base(s) was specified */
-        for (slot = spec->slots; slot->slot; slot++) {
-            if (slot->slot == Py_tp_base)
-                base = slot->pfunc;
-            else if (slot->slot == Py_tp_bases) {
-                bases = slot->pfunc;
-            }
-        }
-        if (!bases) {
-            bases = PyTuple_Pack(1, base);
-            if (!bases)
-                goto fail;
-        }
-        else if (!PyTuple_Check(bases)) {
-            PyErr_SetString(PyExc_SystemError, "Py_tp_bases is not a tuple");
-            goto fail;
-        }
-        else {
-            Py_INCREF(bases);
-        }
+        goto finally;
     }
-    else if (!PyTuple_Check(bases)) {
-        bases = PyTuple_Pack(1, bases);
-        if (!bases)
-            goto fail;
+
+    if (!metaclass) {
+        metaclass = &PyType_Type;
     }
-    else {
-        Py_INCREF(bases);
+    metaclass = _PyType_CalculateMetaclass(metaclass, bases);
+    if (metaclass == NULL) {
+        goto finally;
     }
-
-    PyTypeObject *metatype = _PyType_CalculateMetaclass(&PyType_Type, bases);
-    if (metatype == NULL) {
-        Py_DECREF(bases);
-        goto fail;
+    if (!PyType_Check(metaclass)) {
+        PyErr_Format(PyExc_TypeError,
+                     "Metaclass '%R' is not a subclass of 'type'.",
+                     metaclass);
+        goto finally;
     }
-    res = (PyHeapTypeObject*)metatype->tp_alloc(metatype, nmembers);
+    if (metaclass->tp_new != PyType_Type.tp_new) {
```

Comment:
I am now thinking that this check should probably be:
```
original_metaclass->tp_new != metaclass->tp_new
```
That will (a bit more precise) allow creating an instance of a metaclass with a custom `tp_new`, so long you pass said metaclass explicitly.

Reply:
The problem I see is that we can't really *call* `tp_new` properly, since we're not getting the arguments to pass to it.
We could call it without arguments, but that sounds half-baked.

Do you have a use case for allowing custom `tp_new`?